### ISSUE TITLE
supplementary location services

### ISSUE THREAD

**Issue by thestinger** [GrapheneOS/os_issue_tracker#24]

**Comment by cwmke** [comment]
Have you considered something like Mozilla Location Services?

https://github.com/microg/IchnaeaNlpBackend

**Comment by thestinger** [comment]
I think I'd want it to be local, at least for the main option.

**Comment by lazerl0rd** [comment]
@thestinger there are few good microG UNLP plugins that contain the ability to keep a local db:

https://github.com/microg/OpenWlanMapNlpBackend and https://github.com/wish7code/org.openbmap.unifiedNlpProvider both contain online and offline options.

https://github.com/n76/wifi_backend, https://github.com/rtreffer/LocalGSMLocationProvider, and https://github.com/n76/Local-GSM-Backend are all local-only databases.

https://github.com/microg/android_packages_apps_UnifiedNlp, and of course their respective repos, expands a bit on them more.

**Comment by renlord** [comment]
something like this? https://f-droid.org/en/packages/org.fitchfamily.android.dejavu/

**Comment by OlafPettersson** [comment]
What would best approach here be? Both these plugins to microg's Unified Location Provider can do offline locations:
- https://f-droid.org/en/packages/com.google.android.gms/
- https://f-droid.org/en/packages/org.openbmap.unifiedNlp/

Do I understand correctly, that MicroG will not be included on GrapheneOS? If that's the case, one way forward might be to port microg's Unified Location Provider to GrapheneOS, and then talk to the aforementioned packages (and any others the user might desire to install, e.g. the geocoding package, solving  #23 as well.
Would this be a viable way?

**Comment by thestinger** [comment]
The implementation needs to be a lot more minimal and robust.

**Comment by OlafPettersson** [comment]
Can you elaborate on minimal and robust? To me, from a security standpoint, it seems like a reasonable setup: a higher privileged location provider, and some lower privileged actual implementations.

I can't vouch for the robustness of the above-mentioned plugins. If nessessary they probably can be rewritten more robustly. But what stands against using the same interface to those plugins - or replacements thereof - than microg?
If a rewrite turns out to be desired or neccessary that would also allow to re-use these plugins in microg.

**Comment by Mynacol** [comment]
## Data Sources

Only downloadable databases are desired.
Databases used by existing UnifiedNLP backends are:

| source | cell tower | WiFi | file format | size (approx.) | license |
| :----- | :--------: | :--: | :---------: | :------------: | ------: |
| [openBmap](https://radiocells.org/downloads) | y | y | sqlite | 2.1 GB | Creative Commons Attribution-ShareAlike 3.0 Unported (cellular) and Open Database License (WiFi) |
| [Mozilla Location Services](https://location.services.mozilla.com/downloads) | y | n | [CSV](https://ichnaea.readthedocs.io/en/latest/import_export.html) | 250 MB | Public Domain |
| [OpenCellid](https://opencellid.org/downloads.php) | y | n | [CSV](https://ichnaea.readthedocs.io/en/latest/import_export.html) | 980 MB | CC BY-SA 4.0 |

openBmap includes opencellid data (according to the [license](https://radiocells.org/license) page), and I guess Mozilla data as well. The last update was from 2020-02-16.
Mozilla and OpenCellid data is smaller, only has cell tower data, and is provided as well-defined CSV. The data should probably be converted to sqlite to be indexable.

[FastLacellsGenerator](https://github.com/sobrus/FastLacellsGenerator) combines all three to one big sqlite database to be used by [Local GSM Backend](https://gitlab.com/deveee/Local-GSM-Backend). This could be done as part of the system building process.


## Architecture

I think the plugin-based system of UnifiedNLP is not a bad idea. This way we can easily add support for other location providers (WiFi, Bluetooth beacons). If AOSP already supports multiple (coarse) location providers at the same time we can drop this idea.
Some considerations:
The backend (de-)registration is unneeded. We can hard-depend on the shipped backend(s). This removes some intents, and the other intents can only be allowed by the defined system backend(s). So no (user) app can get new intents/resorces/permissions.
The GUI/Settings and download stuff can be removed _if_ we ship the database with the system image. Then the app focuses on location.


## Database deployment

Bundling the database with system images dramatically increases their size. The main alternative would be a separate download, ideally signed. Then we need a GUI for it, explaining users why they would want it etc.
So for now bundling with the OS is the easiest method.

I assume we would ship data for the whole world, and not selecting specific countries.

**Comment by thestinger** [comment]
A plugin-based model without enforced signature verification bypasses the intended security model. There are also issues with hot this is implemented. It's not going to be implemented that way. We're also definitely not including microG due to severe security issues with it. The only way this will be included is as a minimal and robust app. It should have no reason to be privileged beyond being whitelisted as a location provider. The privilege separation that you're talking about not only doesn't make sense but as implemented breaches the security model by bypassing the whitelist of providers.

**Comment by thestinger** [comment]
Also, please note that anything licensed as GPLv3 or AGPLv3 also cannot be included in GrapheneOS.

The desired approach is implementing a location provider as an app where our releases of it are whitelisted. We want to have one that's included by default and has a simple toggle for enabling it and beginning to keep the database updated.

AOSP already supports whitelisting multiple location providers. Unless there's a good reason not to use the standard AOSP system for this, that's what we'll want to use. We don't want to have a central app that's whitelisted and which needs to implement a whole separate setup for whitelisting plugins. It makes more sense to use the existing system as long as it works properly for the desired implementation.

This should have nothing to do with a compatibility layer for Play services and we're never going to have any kind of support for microG. There will be a stubbed out Play services compatibility layer as a disabled by default option available to users as described in https://grapheneos.org/#never-google-services but that has nothing to do with this. Geocoding as mentioned in #23 also has nothing to do with Play services. It is an AOSP API requiring a provider, just like this, and needs to be implemented in an appropriate way.

**Comment by OlafPettersson** [comment]
From a feature, license and maintenance perspective https://gitlab.com/deveee/Local-GSM-Backend seems to be a reasonable starting point. The implementation supports both mozilla location service and opencellid, downloads the data into a local sqlite db and seems to have a robust location algorithm (in [TelephonyHelper.java](https://gitlab.com/deveee/Local-GSM-Backend/-/blob/master/app/src/main/java/org/fitchfamily/android/gsmlocation/TelephonyHelper.java).
It is licensed under the Apache License and seems to be well maintained.

I'm not sure what would be the best way to integrate the package into GrapheneOS as µG is not an option.

In any case, I'm not sure a single configuration switch will suffice, since the amout of data that needs to be downloaded regulary is quite large. Also, most users probably want to select only a few countries for which to download and update GMS data.

**Comment by OlafPettersson** [comment]
tl;dr: For offline-location based on cell data from opencellid.org and Mozilla Location Service, only a very rough location estimate with accuracy between 500m and 10km is achievable. WiFi data is not available from non-proprietary sources for offline usage. 

To get a better grip on where the open source community stands with network location providers on android, I wrote a simple app, "[Location Tester](https://github.com/OlafPettersson/LocationTester)". It instantiates and calls UnifiedNlp-Plugins by their very simple [API](https://github.com/OlafPettersson/LocationTester/blob/master/app/src/main/aidl/org/microg/nlp/api/LocationBackend.aidl). The app does not need MicroG or their UnifiedNlp-Service installed.

Here are my observations:

Name   | License Source Code | Operation | Remarks
------ | --      | --        | ------
LocalGsmNlpBackend<br>org.fitchfamily.android.gsmlocation | Apache License 2.0 | Offline / GSM/UMTS/LTE | Uses an offline database blended from [opencellid.org](https://opencellid.org) and [Mozilla Location Service (MLS)](https://location.services.mozilla.com/). Can download directly from these services or from a [repository at github](https://github.com/wvengen/lacells). <br> As for the locations, for me, results are mixed. The phones I have available only report the stongest network cell, not neighboring cells. Triangulation is thus impossible. The strongest cell near my home is not mapped by either opencellid or MLS. When in other places, I do get locations from this backend. These locations are often both off and have a quite large reported confidence area of 1000-5000m. The true location is most of the time in the reported confidence area. <br> <br>As a side note, as far as I can see, the returned confidence radius does not take into account the definition differences of Android (68% confidence) with ranges returned from MLS ([some kind of "100%" confidence](https://ichnaea.readthedocs.io/en/latest/algo/observations.html#the-station-model)) or from opencellid ([98% confidence](https://community.opencellid.org/t/documenting-the-columns-in-the-downloadable-cells-database-csv/186)). 
Déjà Vu<br>org.fitchfamily.android.dejavu | GPL3 | Offline / WiFi/GSM/UMTS/LTE | *UnifiedNlp backend that uses locally acquired WLAN/Wi-Fi AP and mobile/cellular tower data to resolve user location. This backend uses no network data. All data acquired by the phone stays on the phone.*<br><br>I found that this gives very good results when able to learn from an enabled GPS. I believe it might be a problem for GrapheneOS that this backend basically builds an on-phone database of visited locations. If you would ever have to hand over your phone-data, you would have to hand over where you have been.
LocalWifiNlpBackend<br>org.fitchfamily.android.wifi_backend | GPL3 | Offline / WiFi | Superseeded by Déjà Vu backend. Works only with WiFi-data. Gives quite good results when able to learn from enabled GPS.
Radiocells.org UnifiedNlp Backend<br>org.openbmap.unifiedNlp | AGPL3 | Offline/Online / WiFi/cell | Radiocells.org seems to be put on [hibernation](https://twitter.com/radiocells/status/1194332874444554247). The latest data update appears to be from January 2020. I could get online mode to work, but not offline mode.
MozillaNlpBackend<br>org.microg.nlp.backend.ichnaea | Apache License 2.0 | Online / WiFi/GSM/UMTS/LTE | You hand your IP, WiFi and cell data to Mozilla and get a location. Can be configured to only use WiFi or cell. For me results are ok to mixed. The reported location/confidence sometimes appears to be wrong/too small when using WiFi. It is large (>1000m) but mostly correct when using cell.<br>Mozilla [does not share WiFi data](https://location.services.mozilla.com/downloads) for offline use.
Apple UnifiedNlp Backend<br>org.microg.nlp.backend.apple | Apache License 2.0  | Online / WiFi  | I did not test this myself. But what I have seen is, when indoors or in developed areas, this gives very accurate results. The reported confidence area appears to be larger than needs to be. Obviously, everytime you use this plugin, you hand over your IP address and location to Apple.


<br>
I am not sure how to go forward with this. I find the GNSS on my Pixel4a quite quick and accurate.  I am not so sure any of the non-proprietary data sources can add much value to it.

**Comment by thestinger** [comment]
Yeah, that doesn't seem very useful, particularly since there's carrier-based SUPL.

**Comment by thestinger** [comment]
GNSS + carrier-based A-GPS is more than good enough in most cases. I don't want to have network-based location services or ones keeping a local database of your movements. The only acceptable approach is a static database covering a large general area like a city. Since it sounds like Wi-Fi databases are not available, the only option is a static cell tower database. Carrier-based A-GPS already offers something along those lines as a way to bootstrap GNSS and should be working perfectly already. It's something we intend to look into in terms of choosing the best fallback server, etc. but it should be fine already. When using airplane mode, it's not like you could make use of the offline database for cell towers. Since we only seem to have a viable path to using cell towers, it won't work for when the cellular radio is disabled or on a tablet without one.

It may improve app compatibility to have support for this since some apps seemingly don't wait long enough for GNSS or can't use it at all. It's quite likely that most of those apps use Play services APIs for it and nothing we do will make them work though. Not clear that there is anything substantial to gain from that angle.

Does anyone else have thoughts on it? Is there a good reason to go forward with implementing a static database using cell towers? I think the time may be better spent experimenting with SUPL and seeing if we can offer a choice of fallback providers when the carrier doesn't use their own implementation.

**Comment by Victor239** [comment]
> static database using cell towers

How is this database going to be any different from [OpenCellID](https://www.opencellid.org/) or [Radiocells/OpenBmap](https://radiocells.org/)?

**Comment by thestinger** [comment]
Please read the thread.

**Comment by Victor239** [comment]
I did and I failed to see a good argument as to why yet another cellular database™ will have any better data than existing ones, only addressed was your arguments against the architecture of the plugins which contain those databases.

**Comment by thestinger** [comment]
I think you're misunderstanding the purpose of the thread and what was being discussed. There has never been any plan to gather our own data for a cellular database.

**Comment by Victor239** [comment]
I see, my apologies then, it seemed as if _both_  the architecture and databases of existing location providers were deemed insufficient which just came across as NIH syndrome.

**Comment by thestinger** [comment]
The data in those databases does appear to be insufficient which means we probably won't be implementing it.

**Comment by Zanthed** [comment]
@Ammako 
If we're talking about something like Night Light, you could simply set the custom schedule times for it to turn on or off.

**Comment by iAnyKey** [comment]
The lack of NLP support was in my case the reason to switch from GrapheneOS to other AOSP-based OS after 1 Year of usage.
To keep the things short:
* pure-GPS Navigation massively limits usability of the smartphone as a primary device. Primarily because of its slow - try to call a taxi being abroad.
* I don't see right now the limitations of the NLP sources mentioned previously. Exactly as @OlafPettersson found out in his tests - network based location accuracy is more than sufficient for daily city life and offers reasonable response times.
* GrapheneOS still feels for me way more secure, up-to-date and private, than competitor.
However sadly I have no idea, how one can contribute to GrepheneOS regarding this aspect.

**Comment by thestinger** [comment]
Complaining isn't an appropriate use of the issue tracker.

**Comment by iAnyKey** [comment]
My apologies. It has been meant that way.

Am 5. Jänner 2022 13:09:57 MEZ schrieb Daniel Micay ***@***.***>:

**Comment by JordanBoulan** [comment]
This would be an extremely nice feature to have. In order to get accurate location on Graphene you are basically forced to turn on the improved location accuracy using Google's online location database. I can't even get maps to find my location without it in a normal suburban area outside a big city.

When I was testing Calyx OS when I first got my pixel 6 the UnifiledNlp originating from lineage os worked quite well using Mozillia and Dejavu backends. 

**Is this not better than sending my location to google servers, as I am forced to do currently in GrapheneOS especially for working navigation? Or am I misunderstanding an implication or the current Graphene implementation?**

Since this is already a feature in Lineage and Calyx, it does not seem like it would be to hard to integrate it in as an optional feature, though I could be wrong.

**Shouldn't users and devs have the choice to use a different location backend than Google?** I think this issue deserves more attention or at least higher priority. It would definitely make a huge difference for me as a user. I am trying to get away from google but still sending them my location for apps and navigation needs.

Thanks a bunch.

**Comment by thestinger** [comment]
Please stop promoting a malicious project (CalyxOS) on our issue tracker. It will result in a ban.

**Comment by thestinger** [comment]
We plan on making a robust and secure implementation using a local cell tower database via the publicly available data for it. GrapheneOS does not take shortcuts seriously compromising user privacy and security like CalyxOS and LineageOS. The insecure implementation from microG breaks the privacy/security model for location and we aren't going to send user location to a service with questionable privacy. A service being hosted by Mozilla doesn't make it inherently private in any way. We aren't going to send the user's location to a service or break the security model. We also aren't going to add historical location tracking which is not a good workaround for lack of a real implementation of this. If you want shortcuts and hacks, this is the wrong project.

**Comment by JordanBoulan** [comment]
@thestinger 

I simply mentioned that it also contained that feature of unifiednlp because that is how I was able to test said feature and that unifiednlp worked well for me. I was not promoting said OS in any way and simply said I tested it. I find it funny that you delete my comment within two minutes over something completely trivial when I spent a good amount of time to type 3 paragraphs strongly supporting the issue you authored. You could have simply edited out the mention you did not like, instead you take away my time and effort promoting your cause for something trivial and heavily weighted by your own opinion and bias.

What a disrespectful and non-productive way to act.  I am not trying to be rude, but constructive. I feel that it was very rude to me to simply delete my long and thoughtful comment that simply mentioned a project you dislike. Just please be more thoughtful. I was in strong support of your issue.

Thank you.

**Comment by thestinger** [comment]
Discussing and especially promoting CalyxOS on our issue tracker is against the rules for our issue tracker. I'm not taking away your time and effort by removing your comment. You're welcome to repost it without breaking that rule. I'm not going to edit other people's comments.

**Comment by thestinger** [comment]
It has nothing to do with bias, and upholding the rules we've decided on is not disrespectful or rude. This is taking time away from the Android 12.1 port for an argument about our rules when it comes to that malicious project.

**Comment by thestinger** [comment]
Also, as you can see, my own comments are deleted too, because it's off-topic and it shouldn't be discussed here.

**Comment by JordanBoulan** [comment]
I guess you have a loose version of promoting. I tried unifiednlp when I was testing out OS's when I first got my phone as I first stated in my comment. However, I am an avid GrapheneOS user and was supporting your issue. I am aware of the issues you have pointed out of said OS, I was merely pointing out I had a tried a feature that was helpful and thought that considering it for GrapheneOS was worth more of a discussion. I was not aware of the rule that merely mentioning their name was banned and that's on me. However, your process isn't conducive to furthering your goals.

I am simply a dev and user of graphene os making their first time poster here saying that I saw a feature there that originated from their parent as discussed above, that might be worth another look. I even originally said the feature (unifiednlp) originated from the parent and not from said OS, it just happened to be where I tried it. Don't try to make it look like I was promoting them, that was simply not the case. Same as you won't edit a few words and kindly remind a first-time poster of the rules, I won't rewrite something I just spent 15 min on.

Yes, I didn't read the rules (I doubt many other's in my position would have either), however your process isn't conducive to growing your community or welcoming to new people, which is maybe something to think about when moderating in the future. I for one won't be posting here again or contributing to the project. 

I heard on reddit when researching OS's that the graphene community was unwelcoming, sad to see that it's unfortunately actually true.

**Comment by thestinger** [comment]
@JordanBoulan You're continuing to break the rules and your attempts to cause drama and attack the project will result in having to ban you from our issue tracker. This is the final warning.

**Comment by thestinger** [comment]
@JordanBoulan Any further attempts to stir up trouble or attacks on the project will result in a ban from our community as a whole including Matrix/Telegram/IRC. It's you who is being toxic and disruptive. You're disrupting the migration to Android 12.1 and are harming the overall GrapheneOS community. There's a reason we moved away from using Reddit due to it being full of misinformation and highly toxic people brigading our community. There's a reason for the toxic and malicious CalyxOS community not being welcome in our community. Years of bullying, harassment, misinformation and attacks on our project have led to this. Participating in that is only going to confirm why the rules are in place and enforced.

You could have reposted your original message with the problematic portion left out and we could have had a discussion about that but you've chosen disruption instead. This thread is being temporarily locked for 24 hours to avoid further disruption.

**Comment by thestinger** [comment]
This issue is going to have to be deleted and replaced with a new one due to an attempt by someone (the one who left a negative reaction on that comment) to brigade our issue tracker again.

**Comment by thestinger** [comment]
To clarify for people coming to see this later on, @JordanBoulan is not to blame for this issue currently being locked but rather someone on Twitter is trying to get people to come bother us based on a misrepresentation of recent activity here.
